File: packages/audio-duration/test/fixtures.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

import { estimateCeilingSeconds, probeDuration, ESTIMATOR } from '../src/index.js';
import { estimateCeilingSecondsFromMultipart } from '../src/index.js';

// Walk up to the repo root rather than counting directories: this file
// runs both from source and from the compiled dist-test tree, which sit
// at different depths.
const FIXTURES = (() => {
  const rel = 'livepeer-network-protocol/extractors/fixtures/multipart-audio-duration-v1';
  let dir = dirname(fileURLToPath(import.meta.url));
  for (let i = 0; i < 8; i++) {
    const candidate = join(dir, rel);
    if (existsSync(join(candidate, 'manifest.json'))) return candidate;
    dir = dirname(dir);
  }
  throw new Error(`shared fixtures not found above ${dirname(fileURLToPath(import.meta.url))}`);
})();

interface Fixture {
  file: string;
  format?: string;
  seconds?: number;
  ceiling_seconds?: number;
  reject: boolean;
  why: string;
}

const manifest = JSON.parse(readFileSync(join(FIXTURES, 'manifest.json'), 'utf8')) as {
  estimator: string;
  rounding: string;
  fixtures: Fixture[];
};

test('manifest describes the estimator this package implements', () => {
  assert.equal(manifest.estimator, ESTIMATOR);
  assert.ok(manifest.fixtures.length > 0, 'manifest declares no fixtures');
});

// The shared fixtures are the contract between this implementation and
// the broker's. A client computes a funding ceiling before the work runs
// and the broker bills from its own parse afterwards; if the two
// disagree, the settlement exceeds the ceiling and a correct exchange is
// refused. Running the SAME files on both sides is what makes that a
// test failure here rather than an incident in production.
for (const f of manifest.fixtures) {
  test(`fixture ${f.file}: ${f.why}`, () => {
    const bytes = new Uint8Array(readFileSync(join(FIXTURES, f.file)));

    if (f.reject) {
      assert.throws(
        () => estimateCeilingSeconds(bytes),
        `accepted a fixture that must be refused: ${f.why}`,
      );
      return;
    }

    const ceiling = estimateCeilingSeconds(bytes);
    assert.equal(ceiling, f.ceiling_seconds, 'ceiling disagrees with the shared manifest');

    const res = probeDuration(bytes);
    assert.equal(res.format, f.format);
    assert.ok(res.exact, 'a funding ceiling cannot be built on an estimate');
    assert.ok(
      Math.abs(res.seconds - (f.seconds ?? 0)) < 1e-6,
      `seconds ${res.seconds} != ${f.seconds}`,
    );
  });
}

test('a multipart upload resolves to its file part', () => {
  const audio = new Uint8Array(readFileSync(join(FIXTURES, 'wav-16k-mono-3s.wav')));
  const boundary = 'testboundary';
  const enc = { encode: (t: string) => Uint8Array.from(t, (c) => c.charCodeAt(0) & 0xff) };
  const head = enc.encode(
    `--${boundary}\r\nContent-Disposition: form-data; name="model"\r\n\r\nwhisper-1\r\n` +
      `--${boundary}\r\nContent-Disposition: form-data; name="file"; filename="a.wav"\r\n` +
      `Content-Type: audio/wav\r\n\r\n`,
  );
  const tail = enc.encode(`\r\n--${boundary}--\r\n`);
  const body = new Uint8Array(head.length + audio.length + tail.length);
  body.set(head, 0);
  body.set(audio, head.length);
  body.set(tail, head.length + audio.length);

  const ceiling = estimateCeilingSecondsFromMultipart(
    body,
    `multipart/form-data; boundary=${boundary}`,
  );
  assert.equal(ceiling, 3);
});

test('a raw body is treated as the audio itself', () => {
  const audio = new Uint8Array(readFileSync(join(FIXTURES, 'wav-16k-mono-3s.wav')));
  assert.equal(estimateCeilingSecondsFromMultipart(audio, 'audio/wav'), 3);
});

test('a missing file part is refused, not defaulted', () => {
  const enc = { encode: (t: string) => Uint8Array.from(t, (c) => c.charCodeAt(0) & 0xff) };
  const body = enc.encode(
    `--b\r\nContent-Disposition: form-data; name="model"\r\n\r\nwhisper-1\r\n--b--\r\n`,
  );
  assert.throws(() => estimateCeilingSecondsFromMultipart(body, 'multipart/form-data; boundary=b'));
});

File: capability-broker/internal/extractors/audioduration/fixtures_test.go
package audioduration

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"testing"
)

const fixtureDir = "../../../../livepeer-network-protocol/extractors/fixtures/multipart-audio-duration-v1"

type fixtureManifest struct {
	Estimator string `json:"estimator"`
	Rounding  string `json:"rounding"`
	Fixtures  []struct {
		File           string  `json:"file"`
		Format         string  `json:"format"`
		Seconds        float64 `json:"seconds"`
		CeilingSeconds int64   `json:"ceiling_seconds"`
		Reject         bool    `json:"reject"`
		Why            string  `json:"why"`
	} `json:"fixtures"`
}

// The shared fixtures are the contract between this implementation and
// any other — a client computing a funding ceiling before the work runs
// must reach the same number this reaches when it bills. Two parsers
// that disagree produce a ceiling the settlement then exceeds, and the
// disagreement surfaces as a refused exchange rather than a bug report.
//
// So this runs the published fixtures rather than its own copies of
// them. A change here that the fixtures do not cover is a change another
// implementation cannot know about.
func TestSharedFixtures(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join(fixtureDir, "manifest.json"))
	if err != nil {
		t.Fatalf("shared fixtures unreadable: %v", err)
	}
	var m fixtureManifest
	if err := json.Unmarshal(raw, &m); err != nil {
		t.Fatal(err)
	}
	if m.Estimator != "multipart-audio-duration/v1" {
		t.Fatalf("manifest estimator = %q", m.Estimator)
	}
	if len(m.Fixtures) == 0 {
		t.Fatal("manifest declares no fixtures")
	}

	for _, f := range m.Fixtures {
		t.Run(f.File, func(t *testing.T) {
			body, err := os.ReadFile(filepath.Join(fixtureDir, f.File))
			if err != nil {
				t.Fatal(err)
			}
			// The ESTIMATOR contract, which is the shared one: exact
			// ceiling or refusal. Probe is the broker-internal view and
			// reports inexact results for the billing path to apply
			// policy to; a funding ceiling has no such latitude.
			ceiling, err := EstimateCeilingSeconds(body)

			if f.Reject {
				if err == nil {
					t.Fatalf("accepted a fixture that must be refused (%s): got %ds",
						f.Why, ceiling)
				}
				return
			}
			if err != nil {
				t.Fatalf("refused a measurable fixture (%s): %v", f.Why, err)
			}
			if ceiling != f.CeilingSeconds {
				t.Fatalf("ceiling = %d; want %d", ceiling, f.CeilingSeconds)
			}
			// The raw duration and format are checked too, so a change
			// that happens to round to the same ceiling still shows up.
			res, perr := Probe(body)
			if perr != nil {
				t.Fatalf("probe: %v", perr)
			}
			if string(res.Format) != f.Format {
				t.Fatalf("format = %q; want %q", res.Format, f.Format)
			}
			if got := res.Duration.Seconds(); math.Abs(got-f.Seconds) > 1e-6 {
				t.Fatalf("seconds = %v; want %v", got, f.Seconds)
			}
		})
	}
}
